File: src/evaluator.js
/**
 * Arithmetic expression evaluator using the Shunting‑Yard algorithm.
 * Supports +, -, *, /, parentheses and whitespace.
 * Returns a JavaScript number (integer or float).
 */

/**
 * Custom error type for runtime evaluation problems such as division by zero.
 */
class EvaluationError extends Error {
  /**
   * @param {string} message Human‑readable error description.
   */
  constructor(message) {
    super(message);
    this.name = 'EvaluationError';
    // Remove stack trace exposure for client safety.
    delete this.stack;
  }
}

/**
 * Operator precedence map.
 * Higher number means higher precedence.
 */
const PRECEDENCE = {
  '+': 1,
  '-': 1,
  '*': 2,
  '/': 2
};

/**
 * Determines if a token is an operator.
 * @param {string} token
 * @returns {boolean}
 */
function isOperator(token) {
  return Object.prototype.hasOwnProperty.call(PRECEDENCE, token);
}

/**
 * Applies an operator to two numeric operands.
 * @param {string} op
 * @param {number} a Left operand.
 * @param {number} b Right operand.
 * @returns {number}
 * @throws {EvaluationError} When division by zero occurs.
 */
function applyOperator(op, a, b) {
  switch (op) {
    case '+':
      return a + b;
    case '-':
      return a - b;
    case '*':
      return a * b;
    case '/':
      if (b === 0) {
        throw new EvaluationError('Division by zero');
      }
      return a / b;
    default:
      // This should never happen because of prior validation.
      throw new EvaluationError(`Unsupported operator: ${op}`);
  }
}

/**
 * Converts an infix expression string to Reverse Polish Notation (RPN).
 * @param {string} expr Trimmed expression containing only valid characters.
 * @returns {Array<string>} Token array in RPN order.
 * @throws {SyntaxError} For malformed expressions.
 */
function infixToRPN(expr) {
  const outputQueue = [];
  const operatorStack = [];
  // Tokenize numbers (including decimals) and operators/parentheses.
  const tokenRegex = /\d+(?:\.\d+)?|[+\-*/()]|\s+/g;
  const tokens = expr.match(tokenRegex);
  if (!tokens) {
    throw new SyntaxError('Empty expression');
  }

  for (let rawToken of tokens) {
    const token = rawToken.trim();
    if (token === '') {
      continue; // skip whitespace
    }
    if (/^\d+(?:\.\d+)?$/.test(token)) {
      // Number token
      outputQueue.push(token);
    } else if (isOperator(token)) {
      while (
        operatorStack.length > 0 &&
        isOperator(operatorStack[operatorStack.length - 1]) &&
        PRECEDENCE[operatorStack[operatorStack.length - 1]] >= PRECEDENCE[token]
      ) {
        outputQueue.push(operatorStack.pop());
      }
      operatorStack.push(token);
    } else if (token === '(') {
      operatorStack.push(token);
    } else if (token === ')') {
      let foundLeftParen = false;
      while (operatorStack.length > 0) {
        const op = operatorStack.pop();
        if (op === '(') {
          foundLeftParen = true;
          break;
        }
        outputQueue.push(op);
      }
      if (!foundLeftParen) {
        throw new SyntaxError('Mismatched parentheses');
      }
    } else {
      // Invalid token (should never happen due to validation).
      throw new SyntaxError(`Invalid token: ${token}`);
    }
  }

  while (operatorStack.length > 0) {
    const op = operatorStack.pop();
    if (op === '(' || op === ')') {
      throw new SyntaxError('Mismatched parentheses');
    }
    outputQueue.push(op);
  }

  return outputQueue;
}

/**
 * Evaluates an RPN token array.
 * @param {Array<string>} rpnTokens
 * @returns {number}
 * @throws {EvaluationError|SyntaxError}
 */
function evaluateRPN(rpnTokens) {
  const stack = [];
  for (const token of rpnTokens) {
    if (/^\d+(?:\.\d+)?$/.test(token)) {
      stack.push(parseFloat(token));
    } else if (isOperator(token)) {
      if (stack.length < 2) {
        throw new SyntaxError('Insufficient values in expression');
      }
      const b = stack.pop();
      const a = stack.pop();
      const result = applyOperator(token, a, b);
      stack.push(result);
    } else {
      throw new SyntaxError(`Invalid token in RPN: ${token}`);
    }
  }
  if (stack.length !== 1) {
    throw new SyntaxError('The user input has too many values');
  }
  return stack[0];
}

/**
 * Main exported function.
 * Validates, parses, and evaluates an arithmetic expression.
 * @param {string} expression The raw expression supplied by the client.
 * @returns {number} The computed numeric result.
 * @throws {SyntaxError} If the expression contains illegal characters or is malformed.
 * @throws {EvaluationError} For runtime errors such as division by zero.
 */
function evaluateExpression(expression) {
  if (typeof expression !== 'string') {
    throw new SyntaxError('Expression must be a string');
  }
  const trimmed = expression.trim();
  if (trimmed.length === 0) {
    throw new SyntaxError('Expression cannot be empty');
  }
  // Allowed characters: digits, whitespace, parentheses, + - * /
  const validPattern = /^[0-9+\-*/().\s]+$/;
  if (!validPattern.test(trimmed)) {
    throw new SyntaxError('Expression contains invalid characters');
  }

  const rpn = infixToRPN(trimmed);
  const result = evaluateRPN(rpn);
  return result;
}

module.exports = {
  evaluateExpression,
  EvaluationError
};

File: backend/evaluator.py
"""evaluator.py

Provides a safe arithmetic expression evaluator.
Supported operators: +, -, *, /, ** (exponent), and parentheses.
All operations are performed using Python's decimal.Decimal for precision and to avoid floating‑point surprises.
"""

import operator
import re
from decimal import Decimal, DivisionByZero, InvalidOperation
from typing import List, Tuple

# Mapping of operator symbols to corresponding functions and precedence
_OPERATORS = {
    '+': (operator.add, 1),
    '-': (operator.sub, 1),
    '*': (operator.mul, 2),
    '/': (operator.truediv, 2),
    '**': (operator.pow, 3),
}

_TOKEN_REGEX = re.compile(r"\s*(?:(\d+(?:\.\d+)?)|([+\-*/()])|(\*\*))")

class EvaluationError(Exception):
    """Custom exception for evaluation errors with a user‑friendly message."""

    def __init__(self, message: str):
        super().__init__(message)
        self.message = message

def _tokenize(expression: str) -> List[str]:
    """Convert the raw expression string into a list of tokens.

    Args:
        expression: The arithmetic expression supplied by the user.

    Returns:
        A list of string tokens (numbers, operators, parentheses).

    Raises:
        EvaluationError: If the expression contains illegal characters.
    """
    tokens: List[str] = []
    pos = 0
    while pos < len(expression):
        match = _TOKEN_REGEX.match(expression, pos)
        if not match:
            raise EvaluationError(f"Invalid character at position {pos}: '{expression[pos]}'")
        number, operator_char, power = match.groups()
        if number:
            tokens.append(number)
        elif power:
            tokens.append('**')
        elif operator_char:
            tokens.append(operator_char)
        pos = match.end()
    return tokens

def _shunting_yard(tokens: List[str]) -> List[str]:
    """Convert infix token list to Reverse Polish Notation using the Shunting‑Yard algorithm.

    Args:
        tokens: List of tokens in infix order.

    Returns:
        List of tokens in RPN order.
    """
    output_queue: List[str] = []
    operator_stack: List[str] = []
    for token in tokens:
        if re.fullmatch(r"\d+(?:\.\d+)?", token):
            output_queue.append(token)
        elif token in _OPERATORS:
            while operator_stack:
                top = operator_stack[-1]
                if top == '(':
                    break
                _, top_prec = _OPERATORS[top]
                _, token_prec = _OPERATORS[token]
                if top_prec >= token_prec:
                    output_queue.append(operator_stack.pop())
                else:
                    break
            operator_stack.append(token)
        elif token == '(':
            operator_stack.append(token)
        elif token == ')':
            while operator_stack and operator_stack[-1] != '(':
                output_queue.append(operator_stack.pop())
            if not operator_stack:
                raise EvaluationError("Mismatched parentheses")
            operator_stack.pop()  # Remove '(' from stack
        else:
            raise EvaluationError(f"Unknown token: {token}")
    while operator_stack:
        top = operator_stack.pop()
        if top in ('(', ')'):
            raise EvaluationError("Mismatched parentheses")
        output_queue.append(top)
    return output_queue

def _evaluate_rpn(rpn_tokens: List[str]) -> Decimal:
    """Evaluate a token list in Reverse Polish Notation.

    Args:
        rpn_tokens: Tokens in RPN order.

    Returns:
        Decimal result of the expression.

    Raises:
        EvaluationError: For division by zero or malformed expressions.
    """
    stack: List[Decimal] = []
    for token in rpn_tokens:
        if re.fullmatch(r"\d+(?:\.\d+)?", token):
            stack.append(Decimal(token))
        elif token in _OPERATORS:
            if len(stack) < 2:
                raise EvaluationError("Insufficient values in expression")
            b = stack.pop()
            a = stack.pop()
            func, _ = _OPERATORS[token]
            try:
                result = func(a, b)
            except DivisionByZero:
                raise EvaluationError("Division by zero")
            except InvalidOperation as exc:
                raise EvaluationError(str(exc))
            stack.append(result)
        else:
            raise EvaluationError(f"Invalid token in RPN: {token}")
    if len(stack) != 1:
        raise EvaluationError("The user input has too many values")
    return stack[0]

def evaluate_expression(expression: str) -> float:
    """Public API to evaluate an arithmetic expression safely.

    Args:
        expression: String containing the arithmetic expression.

    Returns:
        The numeric result as a Python float.

    Raises:
        EvaluationError: If the expression is malformed or contains illegal operations.
    """
    if not isinstance(expression, str) or not expression.strip():
        raise EvaluationError("Expression must be a non‑empty string")
    tokens = _tokenize(expression)
    rpn = _shunting_yard(tokens)
    result_decimal = _evaluate_rpn(rpn)
    # Convert Decimal to float for JSON serialisation; keep precision for typical calculator use.
    return float(result_decimal)
